src/prof_recent.c

```diff
@@ -506,33 +519,106 @@ prof_recent_alloc_dump(tsd_t *tsd, void (*write_cb)(void *, const char *),
 	emitter_init(&emitter, emitter_output_json_compact,
 	    buf_writer_get_write_cb(&buf_writer),
 	    buf_writer_get_cbopaque(&buf_writer));
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
 	emitter_begin(&emitter);
 
 	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
 	prof_recent_alloc_assert_count(tsd);
+	prof_recent_t *dump_dummy = prof_recent_alloc_dummy;
+	assert(dump_dummy != NULL);
+	prof_recent_alloc_dummy = prof_recent_alloc_dummy_reserve;
+	assert(prof_recent_alloc_dummy != NULL);
+	assert(prof_recent_alloc_dummy->next == prof_recent_alloc_dummy);
+	const ssize_t dump_count = prof_recent_alloc_count;
+	prof_recent_alloc_count = 0;
+	const ssize_t dump_max = prof_recent_alloc_max_get(tsd);
+	prof_recent_alloc_assert_count(tsd);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
 
-	/*
-	 * Set prof_recent_alloc_max to 0 so that dumping won't block sampled
-	 * allocations: the allocations can complete but will not be recorded.
-	 */
-	ssize_t max = prof_recent_alloc_max_update(tsd, 0);
+	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize,
+		&dump_max);
+	emitter_json_array_kv_begin(&emitter, "recent_alloc");
 
-	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize, &max);
+	buf_writer_set_manual(&buf_writer);
+	prof_recent_t *node = NULL;
+	prof_recent_t *last = NULL;
+	while (true) {
+		malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		if (node == NULL) {
+			node = dump_dummy->next;
+		}
+		assert(node != NULL);
+		bool need_more_batch = false;
+		while (node != dump_dummy) {
+			buf_writer_commit(&buf_writer);
+			prof_recent_alloc_dump_node(&emitter, node);
+			if (buf_writer_overflow(&buf_writer)) {
```

Comment:
Maybe have the dump_node above return if overflowed?

Reply:
You mean `prof_recent_alloc_dump_node()`? By the design of the buffered writer, it doesn't know whether overflow happened or not, because the buffered writer doesn't propagate any error up. To tell so, `prof_recent_alloc_dump_node()` needs to call `buf_writer_overflow()`, so I'd rather put the call here, so that all buffered writer related logic live together.

Reply:
Agreed. Thanks for explaining!